src/objectExplorer/objectExplorerService.ts

```diff
@@ -56,70 +107,155 @@ export class ObjectExplorerService {
+                const children = result.nodes.map(node => TreeNodeInfo.fromNodeInfo(node, self._currentNode.sessionId,
+                    self._currentNode, self._currentNode.connectionCredentials));
+                self._treeNodeToChildrenMap.set(self._currentNode, children);
+                const expandParams: ExpandParams = {
+                    sessionId: result.sessionId,
+                    nodePath: result.nodePath
+                }
+                for (let key of self._expandParamsToPromiseMap.keys()) {
+                    if (key.sessionId === expandParams.sessionId &&
+                        key.nodePath === expandParams.nodePath) {
+                        let promise = self._expandParamsToPromiseMap.get(key);
+                        return promise.resolve(children);
+                    }
+                }
             }
         };
         return handler;
     }
 
-    private async expandNode(node: TreeNodeInfo, sessionId: string): Promise<boolean> {
+    private async expandNode(node: TreeNodeInfo, sessionId: string, promise: Deferred<TreeNodeInfo[]>): Promise<boolean> {
         const expandParams: ExpandParams = {
             sessionId: sessionId,
             nodePath: node.nodePath
         };
         const response = await this._connectionManager.client.sendRequest(ExpandRequest.type, expandParams);
+        if (promise) {
+            this._expandParamsToPromiseMap.set(expandParams, promise);
+        }
         return response;
     }
 
-    async getChildren(element?: TreeNodeInfo): Promise<TreeNodeInfo[]> {
+    private updateNode(node: TreeNodeInfo): void {
+        for (let rootTreeNode of this._rootTreeNodeArray) {
+            if (rootTreeNode.connectionCredentials === node.connectionCredentials &&
+                rootTreeNode.label === node.label) {
+                    const index = this._rootTreeNodeArray.indexOf(rootTreeNode);
+                    this._rootTreeNodeArray[index] = node;
+                    return;
+            }
+        }
+        this._rootTreeNodeArray.push(node);
+    }
+
+    async getChildren(element?: TreeNodeInfo): Promise<vscode.TreeItem[]> {
         if (element) {
-            if (element !== this.currentNode) {
+            if (element !== this._currentNode) {
                 this._currentNode = element;
             }
-            if (this._treeNodeToChildrenMap.get(this._currentNode)) {
+            // get cached children
+            if (this._treeNodeToChildrenMap.has(this._currentNode)) {
                 return this._treeNodeToChildrenMap.get(this._currentNode);
             } else {
-                // expansion
-                await this.expandNode(element, element.sessionId);
-                return [];
+                // check if session exists
+                if (element.sessionId) {
+                    // clean created session promise
+                    this._sessionIdToPromiseMap.delete(element.sessionId);
```

Comment:
So by deleting these without completing you're going to leave dangling promises all over the place. It would really be better practice to resolve them with just undefined or something as appropriate and make sure the callers awaiting them handle those situations

Reply:
How's it dangling? Where is the case where the promise isn't resolved or rejected?

Reply:
mainController.ts

```
                    let promise = new Deferred<TreeNodeInfo>();
                    await self._objectExplorerProvider.createSession(promise);
                    return promise.then(() => {
                        this._objectExplorerProvider.refresh(undefined)
                    });
```

The .then would never be called and thus the Promise would never be fully resolved. I don't believe this would ever cause an actual runtime issue (haven't investigated your usage of these fully) but regardless we'd currently be leaking promises which is something we should avoid.

---
In reply to: [327304232](https://github.com/microsoft/vscode-mssql/pull/1301#discussion_r327304232) [](ancestors = 327304232)